### PR TITLE
111 graph cycles

### LINKED ISSUE #111: Cyclic Graphs should not be allowed to be made.
Current pipeline validation only checks to make sure the graph is connected.  We are not checking if the graph has a cycle.  If the graph does have a cycle, infinite loops / un-defined behavior occurs while running the pipeline.

We should do this validation in the backend.

There are python libraries for graph analysis that we might want to use. Eg: [https://networkx.org/]( https://networkx.org/)

### PR DESCRIPTION
**Pull request by awtkns**
### Cycle and Connectivity Checking
closes #111

- All cycle and connectivity checking is now done in the backend. I am using the graphing library networkx to do this. 
- I added validators to pipelineUpdate that checks for connectivity. 
- I also added a utitiliy func that converts a pipeline into a directed graph should you want to do that. 
- Removed all frontend graph checking, sorry @KlamChowder1. 
- If there are any errors a message is displayed in the frontend

**schema validation**
```python
class PipelineUpdate(BaseModel):
    pipeline_id: Optional[int]
    nodes: Optional[List[PipelineNodeCreate]] = []
    links: Optional[List[PipelineLinkCreate]] = []

    @root_validator
    def check_pipeline_structure(cls, values):
        graph = DiGraph()
        graph.add_nodes_from([v.node_id for v in values.get('nodes')])
        graph.add_edges_from([(e.from_, e.to) for e in values.get('links')])
        validate_pipeline_structure(graph)

        return values
```

**using networkx**
```python
def validate_pipeline_structure(graph: Union[Graph, DiGraph]) -> bool:
    assert is_connected(graph.to_undirected()), 'The pipeline must be connected.'

    try:
        find_cycle(graph)
    except NetworkXNoCycle:
        pass
    else:
        raise ValueError('The pipeline must not contain a cycle.')

    return True

```

### CODE REVIEW

File: backend/tests/test_models/test_pipelines.py [hunk 1]
```diff
@@ -155,3 +160,34 @@ def test_linux_paths(db):
     assert '\\' not in job.output_path
     assert pathlib.Path(job.input_path).as_posix() == job.input_path
     assert pathlib.Path(job.output_path).as_posix() == job.output_path
+
+
+def test_graphs(db):
```
Comment (asim-shrestha) [inline backend/tests/test_models/test_pipelines.py:165]: could also unit test invalid node creation here?

File: backend/tests/test_routes/test_pipeline.py [hunk 2]
```diff
@@ -65,6 +66,145 @@ def test_create_pipeline(authorization_header, pipeline=schemas.pipeline.Pipelin
...
+
+    url = f'/pipeline/{pipeline["id"]}'
+    response = client.post(url, json=data, headers=authorization_header)
+
+    assert response.status_code == 422
+
+    return response.json()
+
+
+def test_create_pipeline_connected(authorization_header):
```
Comment (asim-shrestha) [inline backend/tests/test_routes/test_pipeline.py:179]: minor but could also valid test pipeline with links here

**Review by asim-shrestha — APPROVED** [review]
Nice stuff 🙏

File: backend/api/schemas/pipeline.py [hunk 3]
```diff
@@ -73,6 +79,22 @@ class PipelineUpdate(BaseModel):
...
+    def check_pipeline_structure(cls, values):
+        graph = DiGraph()
+        graph.add_nodes_from([v.node_id for v in values.get('nodes')])
+        graph.add_edges_from([(e.from_, e.to) for e in values.get('links')])
+        validate_pipeline_structure(graph)
+
+        return values
+
+    @validator('nodes')
+    def unique_node_ids(cls, v):
```
[the remark this answers is not in the document]
Comment (awtkns) [reply backend/api/schemas/pipeline.py:92]: The node_id only ever exists in the frontend for the graph builder

[the remark this answers is not in the document]
Comment (asim-shrestha) [reply backend/api/schemas/pipeline.py:92]: gotcha

File: backend/tests/test_models/test_pipelines.py (hunk 1 above)
Comment (awtkns) [reply backend/tests/test_models/test_pipelines.py:165]: Sure. I was mostly just testing the pipeline structure

Comment (asim-shrestha) [reply backend/tests/test_models/test_pipelines.py:165]: (pretty minor though)

**Pull request merged**